Kafka OAuthBearer authentication fails in native mode: DefaultJwtRetriever cannot be instantiated

[kafka-oauthbearer-native-reproducer.zip](https://github.com/user-attachments/files/25421191/kafka-oauthbearer-native-reproducer.zip)

# Kafka SASL/OAUTHBEARER: `DefaultJwtRetriever` cannot be reflectively instantiated in native mode

## Environment

| | |
|---|---|
| **Quarkus version** | `3.31.4` |
| **Extension** | `quarkus-messaging-kafka` |
| **Java** | 21 |
| **Build tool** | Maven |
| **Mode** | Native (GraalVM) |

---

## Describe the bug

When a Quarkus application uses kafka client with `SASL/OAUTHBEARER` authentication and is compiled to a **native binary**, the application crashes at startup with:

```
Caused by: org.apache.kafka.common.config.ConfigException:
  The class org.apache.kafka.common.security.oauthbearer.DefaultJwtRetriever defined in the
  sasl.oauthbearer.jwt.retriever.class configuration could not be instantiated:
  Could not find a public no-argument constructor for
  org.apache.kafka.common.security.oauthbearer.DefaultJwtRetriever
    at org.apache.kafka.common.security.oauthbearer.internals.secured.ConfigurationUtils.getConfiguredInstance(ConfigurationUtils.java:314)
    at org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginCallbackHandler.configure(OAuthBearerLoginCallbackHandler.java:186)
    at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:67)
    at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:116)
    at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:170)
```

The same application **works correctly in JVM mode** — the only failure there is a network-level `Connection refused` because no real Kafka broker / OAuth provider is reachable, which is expected and proves that reflective instantiation is fine in JVM mode.

---

## Expected behavior

The native binary should start without any reflection/instantiation error for `DefaultJwtRetriever` (and related pluggable `sasl.oauthbearer.*` classes).  
A `Connection refused` at the network level is acceptable; a `ConfigException` caused by a missing reflection registration is not.

---

## Actual behavior

The native binary fails immediately at startup with the `ConfigException` shown above.  
The Kafka client discovers `DefaultJwtRetriever` via `Class.forName()` + reflective no-arg constructor invocation.  
Because these classes are not registered for reflection in the Quarkus native build, GraalVM strips the public no-arg constructors and the instantiation fails.

---

## Steps to reproduce

Attached you can find the reproducer kafka-oauthbearer-native-reproducer.zip

### 1 — build the native binary (skip tests)

```shell
./mvnw clean install -DskipTests -Dnative
```

### 2 — Run the native integration test

```shell
./mvnw failsafe:integration-test -Dit.test="MyMessageIT" -Dnative
```

**Result:** ❌ Application crashes immediately with the `ConfigException` above.

### 3 — Confirm JVM mode works

```shell
./mvnw verify
```

**Result:** ✅ Application starts. Only failure is `Connection refused` (no real broker/IdP) — no reflection error.

---

## Relevant configuration

**`src/main/resources/application.properties`**

```properties
mp.messaging.connector.smallrye-kafka.security.protocol=SASL_PLAINTEXT
mp.messaging.connector.smallrye-kafka.sasl.mechanism=OAUTHBEARER

kafka.sasl.login.callback.handler.class=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginCallbackHandler
kafka.sasl.oauthbearer.token.endpoint.url=https://localhost:8080/<tenant-id>/oauth2/v2.0/token

mp.messaging.connector.smallrye-kafka.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required \
  clientId="team-a-client" \
  clientSecret="team-a-client-secret" ;

quarkus.ssl.native=true
```

---

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
@@ -26,6 +26,7 @@ import org.apache.kafka.common.security.auth.SecurityProtocol;
 import org.apache.kafka.common.security.authenticator.AbstractLogin;
 import org.apache.kafka.common.security.authenticator.DefaultLogin;
 import org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler;
+import org.apache.kafka.common.security.oauthbearer.JwtRetriever;
 import org.apache.kafka.common.security.oauthbearer.OAuthBearerToken;
 import org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerRefreshingLogin;
 import org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerSaslClient;
@@ -254,6 +255,7 @@ public class KafkaProcessor {
         collectImplementors(toRegister, indexBuildItem, ConsumerInterceptor.class);
         collectImplementors(toRegister, indexBuildItem, ProducerInterceptor.class);
         collectImplementors(toRegister, indexBuildItem, MetricsReporter.class);
+        collectImplementors(toRegister, indexBuildItem, JwtRetriever.class);
 
         reflectiveClass.produce(ReflectiveClassBuildItem.builder(OAuthBearerSaslClient.class,
                 OAuthBearerSaslClient.OAuthBearerSaslClientFactory.class,
@@ -294,7 +296,7 @@ public class KafkaProcessor {
 
         for (DotName s : toRegister) {
             reflectiveClass.produce(ReflectiveClassBuildItem.builder(s.toString())
-                    .reason(getClass().getName() + " Jackson and " + Capability.JSONB + " support")
+                    .reason(getClass().getName() + ": Kafka Client instantiation")
                     .build());
         }
 

